DependenciesScanner: report command-line arguments for building pcms explicitly

nkcsgexi opened the pull request:
Clang PCM dependencies are currently implicitly built by the compiler. To allow build-system level to build these PCMs explicitly, we need to report the command line arguments from the compiler to the build system via a JSON file emitted from the dependencies scanner.

This PR combines and emits two sets of Clang compiler arguments: (1) arguments Swift used to construct ClangImporter instance, and (2) arguments collected from running Clang's fast dependency scanner. A high level build system can explicitly schedule PCM building by feeding these arguments to a Swift front-end invocation with an action kind `-emit-pcm`.

For testing purposes, a Swift script is also added to build modules explicitly by using these arguments from the JSON file.

rdar://62612967
rdar://62613017

nkcsgexi committed:
DependenciesScanner: report command-line arguments for building pcm explicitly

Bigcheese commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -927,6 +922,49 @@ ClangImporter::create(ASTContext &ctx, const ClangImporterOptions &importerOpts,
+ClangImporter::createClangInvocation(ClangImporter *importer,
+                                     const ClangImporterOptions &importerOpts,
+                                     ArrayRef<std::string> invocationArgStrs,
+                                     std::vector<std::string> &CC1Args) {
```
You should probably make this optional as you have clients that are uninterested in it.

nkcsgexi replied:
That's a good idea. Fixed.

Bigcheese commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -204,14 +204,29 @@ static void recordModuleDependencies(
+    for(auto arg: clangModuleDep.NonPathCommandLine) {
+      allArgs.push_back("-Xclang");
+      allArgs.push_back(arg);
+    }
```
The arguments in `NonPathCommandLine` are intended to be directly appended to a `-cc1` command line, not to the driver.

Bigcheese commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -204,14 +204,29 @@ static void recordModuleDependencies(
+    swiftArgs.reserve(cc1Args.size() * 3);
+    for (auto arg: cc1Args) {
+      swiftArgs.push_back("-Xcc");
+      swiftArgs.push_back("-Xclang");
```
I'm not sure how `-Xcc` works in the Swift driver, but given this is a `-cc1` command line you can't use `-Xclang`, as that's for forwarding driver options to the frontend.

Bigcheese reviewed (changes requested):
This also needs to handle the other stuff that `getFullCommandLine` does for `-fmodule-map-file=` and `-fmodule-file=`.

Bigcheese commented:
> This also needs to handle the other stuff that `getFullCommandLine` does for `-fmodule-map-file=` and `-fmodule-file=`.

Actually, I guess this would be handled at a higher level.

nkcsgexi commented:
Yeah, I think `-fmodule-map-file=` and `-fmodule-file=` are directly handled by Swift in `emitPrecompiledModule` function.

nkcsgexi committed:
test: split ModuleDependencyGraph into graph and action

nkcsgexi committed:
Test: add a script for building modules by using command line arguments emitted from dependency JSON

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci Please smoke test OS X platform

nkcsgexi commented:
@swift-ci Please smoke test OS X  platform

nkcsgexi commented:
@swift-ci Please smoke test OS X platform

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please test

DougGregor commented:
> rdar://62612967
> rdar://62613017

Please replace this with a full description of which this pull request does. The radar numbers are good for automation, and can be at the end, but the pull request description should be understandable to anyone who cannot follow that link.

nkcsgexi commented:
> > rdar://62612967
> > rdar://62613017
> 
> Please replace this with a full description of which this pull request does. The radar numbers are good for automation, and can be at the end, but the pull request description should be understandable to anyone who cannot follow that link.

Good point. I've updated the PR description.

nkcsgexi committed:
[test] DependenciesScanner: running shell commands via a python script instead from Swift

Running shell commands using Swift in simulator tests is hard. We change the mechanism so that
BuildModulesFromGraph.swift prints command line arguments and a simple python script
picks these arguments and actually runs the command.

nkcsgexi commented:
@swift-ci please test

nkcsgexi commented:
@swift-ci please smoke test
